### DISCUSSION TITLE
`Alpine.store` uses getter, therefore usage within an effect erroneously subscribes the effect to that store

### DISCUSSION THREAD

**Discussion by docweirdo** [alpinejs/alpine#3315] [7. Bugs (archived)]
Depending on the current URL I want to make an API request. I create an effect subscribing to a reactive object called `Alpine.$router.path` using [this library](https://github.com/shaunlee/alpinejs-router) which I can recommend by the way.

Before I make the API request I want to already initialise the store with an empty string, since my UI depends on the content of the store. Populating the store after the API request leads to errors because my store is undefined.

This is a minimal reproducible example of my JS:

``` javascript 
document.addEventListener('alpine:init', () => {

  Alpine.effect(async () => {
    let path = Alpine.$router.path

    console.log("path changed to: " + path)

    Alpine.store('someContent', '')
    let data = await fetchData(path);
    Alpine.store('someContent', JSON.stringify(data))
  })
})

```

Now the problem is that setting the value of `someContent` twice seems to trigger the effect again, my console gets spammed with logs of the unchanged path indefinitely. 

I am not sure why this happens, since I at no point get a value from the `someContent` store, I only set the value twice, which shouldn't lead to my effect subscribing to said value. If I remove either `Alpine.store` statement or use a local object made reactive with `Alpine.reactive()`, there is no loop.

I realise I could prevent this behaviour by initialising the store outside of the effect, but that's beside the point. Speaking of besides the point, is there no way to have a method similar to Vue's `watch` in JS which only watches one specific reactive item? Wrapping the last three statements in `queueMicrotask()` does prevent the loop, but then again, the store is not initialised when the UI renders for the first time.

Any tips as to what I am doing wrong or understanding incorrectly are very appreciated.

**Comment by SimoTod** [comment]
I suspect it's because of the way Alpine.store is implemented:
```js
export function store(name, value) {
    if (! isReactive) { stores = reactive(stores); isReactive = true; }

    if (value === undefined) {
        return stores[name]
    }

    stores[name] = value

    if (typeof value === 'object' && value !== null && value.hasOwnProperty('init') && typeof value.init === 'function') {
        stores[name].init()
    }

    initInterceptors(stores[name])
}
```
The list line is actually a getter. 

Not sure how the but could be fixed though.

Re your code, I suspect it will work correctly with then

```js
document.addEventListener('alpine:init', () => {

  Alpine.effect(async () => {
    let path = Alpine.$router.path

    console.log("path changed to: " + path)

    Alpine.store('someContent', '')
    let data = fetchData(path).then((data) => Alpine.store('someContent', JSON.stringify(data))) // FetchData returns a promise so the thennable runs after the effect callback returns
  })
})
```

**Reply by docweirdo** [reply]
Thank you for your answer. I did look through the code for `Alpine.store` but missed that the getter is probably triggered in the last line.

>Not sure how the but could be fixed though.

Couldn't `initInterceptors` just be called with value, instead of using `stores[name]`, which triggers the getter but provides the same object? Or does that function need a reactive object?

> Re your code, I suspect it will work correctly with then

Tried this, it does not make a difference. The Microtask resulting from `.then` is apparently still run before the end of the effect.

**Comment by docweirdo** [comment]
Just encountered the same issue again with some other code. 

Every store's value I set in the effect becomes a dependency, so when I, at some other point in time, try to set that store's value, the whole effect gets triggered. 

There has got to be a different way to implement `initInterceptors`.

**Reply by SimoTod** [reply]
Ah, right. It's because the first call is still tracking the object, true so the second call, even if it's outside the effect, it will trigger it again. 

I believe initInterceptor needs the reactive object but I'm not 100 % sure. I'll have a look tomorrow

**Reply by SimoTod** [reply]
I suppose it works correctly if you define a classic store outside the effect rather than a single value store, right?
```js
Alpine.store('mystore', {
  content: ''
})
Alpine.effect(async () => {
  let path = Alpine.$router.path console.log("path changed to: " + path)
  Alpine.store('mystore').content = ''
  let data = await fetchData(path)
  Alpine.store('mystore').content = JSON.stringify(data)
})
```
I wonder if it shoukd skip initInterceptor when value is not an object, I don't think the code does anything in that case anyway.

**Reply by docweirdo** [reply]
I'll test using an object tomorrow, thanks for the suggestion. 

What I found out just now is that setting the store in a promise using `.then` or `await` does not seem to add the store as a dependency to the effect. E.g.:
``` js
document.addEventListener('alpine:init', () => {
  Alpine.store('someContent', '')

  Alpine.effect(async () => {
    let path = Alpine.$router.path

    console.log("path changed to: " + path)

    let data = await fetchData(path)
    Alpine.store('someContent', JSON.stringify(data))
  })
})
```

``` html
<body x-data>
    <p>Data is: <span x-data x-text="$store.someContent"></span></p>
    <button @click="$store.someContent = (Math.random() + 1).toString(36).substring(7);">Click</button>
  </body>
```
Clicking the Button and updating the store does in this case not trigger the effect. However, when the store is set within the normal code flow of the effect, clicking the button does trigger the effect, e.g.:
``` js
document.addEventListener('alpine:init', () => {
  Alpine.store('someContent', '')

  Alpine.effect(() => {
    let path = Alpine.$router.path

    console.log("path changed to: " + path)

    let data = fetchData(path).then(data => {})

    Alpine.store('someContent', 'dummy')
  })
})
```

**Edit:** Okay I did test the latter example using an object for the store just now and it does not trigger the effect anymore. Is that because the setter is not called anymore by `initInterceptor`, but rather the reactive object is retrieved?

**Reply by docweirdo** [reply]
@SimoTod any news on this?

**Reply by SimoTod** [reply]
@docweirdo news about?

**Reply by docweirdo** [reply]
> I wonder if it shoukd skip initInterceptor when value is not an object, I don't think the code does anything in that case anyway.

I was under the impression that this might be an issue that could be fixed.

Also, in the edit of my last reply I wanted to confirm with you the reason why the issue does not present itself on objects

**Reply by SimoTod** [reply]
I suppose it can be fixed. I'm not sure about the best way, it will need some testing & trial. Feel free to experiment.
Just to clarify, I'm not the maintainer nor part of the team but the best way to get it fixed is to raise a PR.

The reason why the object doesn't break is because of the code above. When you create a store, it sets and, also, gets/tracks `stores[name]`. When you use single-property stores, you are destroying and recreating the store every time so we are actually reassigning `stores[name]` triggering the reactivity.
When you use an object, stores[name] is still tracked but you are changing a subproperty but not reassigning `stores[name]` so it doesn't trigger the reactivity.

**Comment by ekwoka** [comment]
The issue in the title is the use of the effect causing unintended subscriptions.

You can wrap the stuff you don't want subscribed in a `queueMicrotask` callback which would make it execute after the effect is no longer being tracked.

**Reply by docweirdo** [reply]
As I've explained in the description of the issue, wrapping the statements in question in a microtask lead to their execution after the UI is rendered, the prevention of which is the reason to why I have to set them twice in the first place.

In addition to that, these unintended subscriptions should *in no case* arise from setting a reactive variable, only from getting one.
